python/ray/external_storage.py

```diff
@@ -90,6 +129,223 @@ def restore_spilled_objects(self, keys):
+    """The external storage class implemented by smart_open.
+    (https://github.com/RaRe-Technologies/smart_open)
+
+    Args:
+        bucket_name(str): Name of the bucket to store objects.
+        prefix(str): Prefix of objects that are stored.
+        override_transport_params(dict): Overriding the default value of
+            transport_params for smart-open library.
+        track_tail_latency(bool): If True, get_stats will return tail latency.
+            Tracking tail latency requires more memory.
+
+    Raises:
+        ExternalStorageConfigInvalidError: If it fails to setup a S3 client.
+            For example, if boto3 is not downloaded.
+    """
+
+    def __init__(self,
+                 bucket_name: str,
+                 prefix: str = "ray_spilled_object_",
+                 override_transport_params: dict = None,
+                 track_tail_latency: bool = False):
+        try:
+            import boto3
+            from smart_open import open  # noqa
+        except ModuleNotFoundError:
+            raise ExternalStorageConfigInvalidError(
+                "S3 storage is chosen to be a object spilling "
+                "external storage, but boto3 and smart_open[s3] "
+                "is not downloaded.")
+
+        self.bucket_name = bucket_name
+        self.prefix = prefix
+        self.override_transport_params = override_transport_params or {}
+        self.track_tail_latency = track_tail_latency
+
+        self.session = boto3.Session()
+        self.transport_params = {
+            "session": self.session
+        }.update(self.override_transport_params)
+
+        # --validation--
+        s3 = self.session.resource("s3")
+        # Make sure the bucket exists.
+        bucket = s3.Bucket(self.bucket_name)
+        if not bucket.creation_date:
+            raise ExternalStorageConfigInvalidError(
+                f"Bucket name {self.bucket_name} doesn't "
+                "exist or is not reachable.")
+
+        # --stats--
+        # SANG-TODO Implement it for benchmark.
+        self.download_count = 0
+        self.upload_count = 0
+        self.download_avg_latency = 0
+        self.upload_avg_latency = 0
+
+        if self.track_tail_latency:
+            self.download_latencies = []
+            self.upload_latencies = []
+
+    def spill_objects(self, object_refs):
+        keys = []
+        ray_object_pairs = self._get_objects_from_store(object_refs)
+        for ref, (buf, metadata) in zip(object_refs, ray_object_pairs):
+            key = self.prefix + ref.hex()
+            self._spill_object(key, ref, buf, metadata)
+            keys.append(key.encode())
+        return keys
+
+    def restore_spilled_objects(self, keys):
+        for k in keys:
+            key = k.decode()
+            ref = ray.ObjectRef(bytes.fromhex(key[len(self.prefix):]))
+            self._restore_spilled_object(key, ref)
+
+    def get_stats(self):
+        # TODO(sang): Implement it for benchmark.
+        return {}
```

Comment:
Let's remove stats tracking for now; it is fine to track in the benchmark script.